src/gfnff/gfnff_setup.f90

```diff
@@ -0,0 +1,188 @@
+! MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+! GNU Lesser General Public License for more details.
+!
+! You should have received a copy of the GNU Lesser General Public License
+! along with xtb.  If not, see <https://www.gnu.org/licenses/>.
+
+subroutine gfnff_setup(verbose,restart,mol,p_ext_gfnff)
+  use iso_fortran_env
+  use xtb_restart
+  use xtb_type_molecule
+  use gff_param
+  use xtb_setparam, only : ichrg
+  implicit none
+! Dummy
+  !integer,intent(in) :: ich
+  integer,intent(in) :: p_ext_gfnff
+  logical,intent(in) :: restart
+  logical,intent(in) :: verbose
+  type(TMolecule)  :: mol
+! Stack
+  logical            :: ex
+  logical            :: success
+
+  call gfnff_input(mol)
+  call gfnff_set_param(mol%n)
+  if (restart) then
+     inquire(file='gfnff_topo', exist=ex)
+     if (ex) then
+       call read_restart_gff('gfnff_topo',mol%n,p_ext_gfnff,success,.true.)
+       !hbrefgeo is usually set within gfnff_ini2/gfnff_hbset0 equal to initial xyz
+       hbrefgeo=mol%xyz
+       if (success) write(*,'(10x,"GFN-FF topology read from file successfully!")')
+       if (.not.success) then
+          write(*,'(10x,"GFN-FF topology read in did not work!")')
+          write(*,'(10x,"Generating new topology file!")')
+          call gfnff_ini(verbose,ini,mol,ichrg)
+          call write_restart_gff('gfnff_topo',mol%n,p_ext_gfnff)
+       end if
+     else
+       call gfnff_ini(verbose,ini,mol,ichrg)
+       if (.not.mol%struc%two_dimensional) then
+          call write_restart_gff('gfnff_topo',mol%n,p_ext_gfnff)
+       end if
+     end if
+  else if (.not.restart) then
+     call gfnff_ini(verbose,ini,mol,ichrg)
+     call write_restart_gff('gfnff_topo',mol%n,p_ext_gfnff)
+  end if
+
+end subroutine gfnff_setup
+
+subroutine gfnff_input(mol)
+  use iso_fortran_env, only : wp => real64
+  use xtb_type_molecule
+  use gff_param
+  use xtb_setparam, only : ichrg
+  implicit none
+! Dummy
+  type(TMolecule),intent(in) :: mol
+! Stack
+  integer           :: i,j,k
+  integer           :: ni
+  integer           :: ns
+  integer           :: nf
+  integer           :: ich
+  integer           :: iatom
+  integer           :: iresidue
+  integer           :: ifrag
+  integer           :: ibond
+  integer           :: bond_ij(2)
+  real(wp)          :: r
+  real(wp)          :: dum1
+  real(wp)          :: floats(10)
+  logical           :: ex
+  character(len=80) :: atmp
+  character(len=80) :: s(10)
+
+  if (.not.allocated(nb))       allocate( nb(20,mol%n), source = 0 )
+  if (.not.allocated(qfrag))    allocate( qfrag(mol%n), source = 0.0d0 )
+  if (.not.allocated(fraglist)) allocate( fraglist(mol%n), source = 0 )
+  if (.not.allocated(q))        allocate( q(mol%n), source = 0.0d0 )
+
+  if (allocated(mol%pdb)) then
+    read_file_type = 2
+    ini = .true.
+  else if (allocated(mol%sdf)) then
+    read_file_type = 1
+    ini = .false.
+  else
+    read_file_type = 0
+    ini = .true.
+  end if
```

Comment:
We already have an enumerator this purpose in `xtb_mctc_filetypes` and `mol%ftype`.

Reply:
Now using `mol%ftype` and fileType from `xtb_mctc_filetypes`.